### ISSUE TITLE
What should the pragma be named?

### ISSUE THREAD

**Issue by domenic** [domenic/proposal-function-prototype-tostring-censorship#3]
I currently have specced `"use no Function.prototype.toString"`.

This might be fine. It's pretty clear on what it does.

Any other suggestions?

**Comment by ljharb** [comment]
`"use native Function.prototype.toString"`?

**Comment by michaelficarra** [comment]
`"delete Function.prototype.toString"`

Take it out of quotes for extra umph.

**Comment by isiahmeadows** [comment]
How about any of these? They don't imply we're dropping or otherwise aliasing `Function.prototype.toString`, and who says we're bound to the `"use ${foo}"` format anyways (the spec doesn't)?

- `"censor source strings"`
- `"censor sources"`
- `"censor as native"`
- `"ignore sources"`
- `"use no sources"` if you prefer to stick with the `"use ${foo}"` format.

**Comment by ExE-Boss** [comment]
`"censor toString"` could also work.

**Comment by domenic** [comment]
Current spec has been switched (per #17) to `"hide implementation"`. My only concern there is that it might be over-broad, implying e.g. that all public fields will automatically become private fields or something. But, it's pretty good, and I'd be happy keeping it.

**Comment by Alhadis** [comment]
@domenic  What about `"hide source"`?

**Comment by isiahmeadows** [comment]
I like that better than [any of my previous suggestions](https://github.com/tc39-transfer/proposal-function-implementation-hiding/issues/3#issuecomment-391832378). Nice, succinct, direct, and obvious.

(Not TC39, so I have no real power in this. Just a JS dev who likes to observe these discussions.)

Off-topic thought: should the binary JS proposal integrate with this somehow? Not saying they need to merge, but I feel these two could probably provide the same answer just to simplify implementation.

**Comment by ilias-t** [comment]
I'd prefer something shorter like "hide source" or "hide code" to "hide implementation", which is more likely to be misspelled.

**Comment by michaelficarra** [comment]
Still open to ideas here, but right now we are going with `"sensitive"`, `"hide implementation"`, and (if we must) `"preserve implementation"`.

**Comment by Alhadis** [comment]
... again, what's wrong with something short and obvious like `"hide source"`? Honestly, that's all you're really doing.

A function's `name` and `length` are technically part of its "implementation", so the word itself isn't entirely correct here. As for `"sensitive"`, that doesn't explain what it's doing, only hinting at secrecy. It could be interpreted any number of ways.

**Comment by michaelficarra** [comment]
@Alhadis The function name and position information are also hidden within stack traces by the `"hide implementation"` directive.

**Comment by bathos** [comment]
> FYI I would still like feedback on alternative directive names (https://github.com/tc39/ecma262/pull/1739#issuecomment-542967589)

1. Right now, the directive strings don’t communicate their relationship to one another. I think it should be clearer that the effects of what’s being called ‘sensitive’ are a superset of the effects of what’s being called ‘hide implementation’.
2. Right now, the directive strings are inconsistent in terms of (English) grammar. ‘Hide implementation’ is an instruction, but ‘sensitive’ is a description. Directives should always be written as instructions IMO, and notably, the only precedent for this is written as an instruction (‘use strict’).
3. The word ‘implementation’ has an overly broad scope and is often used to mean different things depending on context. I think it should be more clear that it concerns the exposure of source text.
4. The word ‘sensitive’ has this property also — but even more. Because it does more than one thing, I understand why one might reach for a ‘what it’s intended for’ term instead of a ‘what it does’ term — I assume it’s desirable to keep the strings brief and memorable — but all of the things it does do still concern ‘hiding’ or ‘redacting’. When we combine that fact with (1) above, it hints at two alternatives.

I am picturing:

- "use strict"
- "redact source"
- "redact all"

or:

- "use strict"
- "hide source"
- "hide all"

(I have a minor preference for ‘redact’ over ‘hide’, but I think that’s pretty much entirely subjective and I can’t provide a clear justification as with the considerations listed above. I included "use strict" in these lists to help show how they benefit from following a consistent grammatical pattern.)

**Comment by ExE-Boss** [comment]
I agree, `"redact source"` should only hide the source reported by `function.toString()`, but keep `function.name` and `function.length`, which are properties present on the instance.

**Comment by devsnek** [comment]
how about `'hide source'` and `'hide calls'`

**Comment by bathos** [comment]
@ExE-Boss My impression/understanding is that those aspects of the proposal have been resolved (and match what you’ve said).

@devsnek I think those are an improvement, but 'hide calls' fails to communicate that it also hides source.

**Comment by Alhadis** [comment]
`obfuscate` works, and it's even shorter than my original suggestion:

```js
"obfuscate";
"hide source";
```

Bear in mind that even with stack traces involved, `hide source` is still appropriate because you're essentially hiding the "source" (origin) of a call-site.

**Comment by bathos** [comment]
I like the one verb approach, but ‘obfuscate’ has a specific meaning which is not the same as what this directive does. To obfuscate something is not to hide or remove it — it is to make something (which is accessible) too confusing, byzantine, or difficult to understand. The term ‘obfuscated code’ is already common, and reflects those connotations.

**Comment by ljharb** [comment]
What about?:

To hide just the source: `”hide implementation source;”` (this hides line/col info from stacks as well, but not the frames themselves)

To hide the function from stack frames: `”hide implementation from stack;”` (but toString still works)

To hide from both: `”hide implementation”`

Thoughts?

**Comment by bathos** [comment]
@ljharb I like those; they’re clear. I would have figured the preference for brevity would be strong among interested folks, but personally, if choosing between something brief but ??? and these, I’d prefer these.

Re: the second item there, my impression was that there was no such directive. Is that wrong / has that changed?

**Comment by michaelficarra** [comment]
@bathos You are correct and that is a design change I'm not interested in making. We were pretty thorough in the analysis of the use cases we want to support. We presented this 2-directive solution to committee with general agreement (sans @erights).

**Comment by ljharb** [comment]
I’m not intending to be pushing the design change, more that the pragmas might be more understandable that way.

**Comment by hax** [comment]
> To hide just the source: `”hide implementation source;”` (this hides line/col info from stacks as well, but not the frames themselves)
> 
> To hide the function from stack frames: `”hide implementation from stack;”` (but toString still works)
> 
> To hide from both: `”hide implementation”`
> 
> Thoughts?

I'd like to have shorter and simpler directives like:
```js
'hide source';

'hide stack';

'hide source, stack';
```

**Comment by bathos** [comment]
Those seem pretty reasonable to me too @hax, though I’d be slightly concerned by how easy it is to forget the order (`"hide source, stack"` vs `"hide stack, source"`). I’d also point out again that `"hide stack"` on its lonesome isn’t a thing.

**Comment by ljharb** [comment]
the order is exactly why i made my suggestions quasi-hierarchical.

**Comment by hax** [comment]
If we really want `"hide stack" // toString still work`, I think we could allow both order (`"hide source, stack"` and `"hide stack, source"`).

**Comment by devsnek** [comment]
i don't think a pragma should be considered somewhere to place a list, ordered or otherwise. they are raw strings, they shouldn't contain structured data.

**Comment by ljharb** [comment]
@devsnek i agree; my suggestions were conceptually hierarchical but not structured data.

**Comment by hax** [comment]
Well, even we don't support list in a single directive string, it could be a list anyway:

```js
'use strict';
'sensitive';
'hide implementation';
...
```

Any order can work, and `'hide implementation'` is redundant but won't cause error (as I understand).

**Comment by erights** [comment]
"hide source" rather than "hide implementation"

Pun includes the two things hidden: the source text, and where it is from (url and line:column)

**Comment by michaelficarra** [comment]
I've changed `"hide implementation"` to `"hide source"`. Thoughts on changing `"sensitive"` to `"confidential"` since confidentiality of its source text and local bindings is the security property we will continuously maintain?

**Comment by bathos** [comment]
At first this struck me as an off-kilter word choice due to carrying strong connotations which don’t apply in context (what is _confidential_ is the secrets which have been _confided_ between _confidants_). However many words need to get repurposed a bit for the sake of loose analogies in programming and this usually happens without trouble. It is an improvement over `sensitive` .

Although I don’t think `confidential` is a bad word choice, I do feel strongly that directive\* names should continue to actually be directives\*\*. Both "hide source" and "use strict" are directives\*\*, while "confidential" — on its own — is just an adjective.

\* As in, the thing called a "directive" in ES
\*\* As in its general sense in English (i.e., an instruction)

**Comment by ljharb** [comment]
I agree that a directive needs to *direct* something (transitive verb).

**Comment by erights** [comment]
Currently, is the only operational difference is that `"sensitive";` also hides the stack frame. What if we made it do only that and called it `"hide frame";`? When I think about the reasons I might want to hide sources, vs the reasons I might want to hide frames, I think they're mostly independent.

**Comment by rbutera** [comment]
`"private source"`
